packages/@aws-cdk/aws-apigatewayv2/lib/api.ts

```diff
@@ -0,0 +1,522 @@
+   */
+  readonly failOnWarnings?: boolean;
+}
+
+/**
+ * Represents a WebSocket API in Amazon API Gateway v2.
+ *
+ * Use `addModel` and `addLambdaIntegration` to configure the API model.
+ *
+ * By default, the API will automatically be deployed and accessible from a
+ * public endpoint.
+ */
+export class Api extends Resource implements IApi {
+
+  /**
+   * Creates a new imported API
+   *
+   * @param scope scope of this imported resource
+   * @param id identifier of the resource
+   * @param apiId Identifier of the API
+   */
+  public static fromApiId(scope: Construct, id: string, apiId: string): IApi {
+    class Import extends Resource implements IApi {
+      public readonly apiId = apiId;
+    }
+
+    return new Import(scope, id);
+  }
+
+  /**
+   * The ID of this API Gateway Api.
+   */
+  public readonly apiId: string;
+
+  /**
+   * API Gateway stage that points to the latest deployment (if defined).
+   */
+  public deploymentStage?: Stage;
+
+  protected resource: CfnApi;
+  protected deployment?: Deployment;
+
+  constructor(scope: Construct, id: string, props?: ApiProps) {
+    if (props === undefined) {
+      props = {};
+    }
+    super(scope, id, {
+      physicalName: props.apiName || id,
+    });
+
+    if (props.protocolType === undefined && props.body === null && props.bodyS3Location === null) {
+      throw new Error("You must specify a protocol type, or import an Open API definition (directly or from S3)");
+    }
+
+    switch (props.protocolType) {
+      case ProtocolType.WEBSOCKET: {
+        if (props.basePath !== undefined) {
+          throw new Error('"basePath" is only supported with HTTP APIs');
+        }
+        if (props.body !== undefined) {
+          throw new Error('"body" is only supported with HTTP APIs');
+        }
+        if (props.bodyS3Location !== undefined) {
+          throw new Error('"bodyS3Location" is only supported with HTTP APIs');
+        }
+        if (props.corsConfiguration !== undefined) {
+          throw new Error('"corsConfiguration" is only supported with HTTP APIs');
+        }
+        if (props.routeSelectionExpression === undefined) {
+          throw new Error('"routeSelectionExpression" is required for Web Socket APIs');
+        }
+        break;
+      }
+      case ProtocolType.HTTP:
+      case undefined: {
+        if (props.apiKeySelectionExpression !== undefined) {
+          throw new Error('"apiKeySelectionExpression" is only supported with Web Socket APIs');
+        }
+        if (props.disableSchemaValidation !== undefined) {
+          throw new Error('"disableSchemaValidation" is only supported with Web Socket APIs');
+        }
+        if (props.routeSelectionExpression !== undefined && props.apiKeySelectionExpression !== KnownRouteSelectionExpression.METHOD_PATH) {
+          throw new Error('"routeSelectionExpression" has a single supported value for HTTP APIs: "${request.method} ${request.path}"');
+        }
+        break;
+      }
+    }
```

Comment:
We have another open PR for APIGWv2 support. Take a look at some API suggestions on how this could look instead - https://github.com/aws/aws-cdk/pull/6432#discussion_r397324722.

Let me know what you think.

Reply:
As a potential consumer of this I don't care if it is exposed via 2 different L2 APIs as long as I can use the functionality of both and get compile time errors if I have a mismatch between configurations and usage. If two different L2 APIs make that easier then I say go for it!

Really hoping to see this feature soon as the v2 APIGW APIs have some great features I want to use!

Reply:
i really like @pahud 's view, will use his view and combine with this PR